language/move-prover/spec-lang/src/translate.rs

```diff
@@ -960,24 +987,38 @@ impl<'env, 'translator> ModuleTranslator<'env, 'translator> {
                     );
                 }
             }
```

Comment:
I guess the assumes/assert cannot use old and result_i. Therefore they are type checked similar to preconditions.

However, the assume/asserts may want to access local variables. This actually makes the type checking business here quite more complicated then until now. To do this right, we would need to check and compile those conditions within the context of the move function, not just its parameters.

To make progress, you may want to leave the full solution for a followup PR. That would do:

- move the condition checking for spec blocks in expressions out of this function and in its own function.
- traverse the expansion AST of a function and evaluate (and partially type check! Because of something like `let x = foo(); spec { assume x > 0 }; ...` requires to know the type of `x`) the spec block conditions in there.

Reply:
In Boogie, asserts and assumes inside a Boogie procedure can use old.  But we don't have to allow that.  We will see if we need that.  So, I agree we should typecheck assert/assume like preconditions.